### PR TITLE
feat: global `prank`

### LINKED ISSUE #1010: Global version of `prank`
With the `CheatTarget` interface
Reference implementation of `warp`: #972

### PR DESCRIPTION
**Pull request by bllu404**
Builds off of #972 and resolves #1010.

### CODE REVIEW

**Comment by bllu404** [comment]
Weird, the failing test passes locally.

File: crates/cheatnet/src/cheatcodes/prank.rs [hunk 1]
```diff
@@ -1,17 +1,45 @@
...
+                self.pranked_contracts
+                    .insert(contract_address, CheatStatus::Cheated(caller_address));
+            }
+            CheatTarget::Multiple(contract_addresses) => {
+                for contract_address in contract_addresses {
+                    self.pranked_contracts
+                        .insert(contract_address, CheatStatus::Cheated(caller_address));
+                }
+            }
+        }
```
Comment (piotmag769) [inline crates/cheatnet/src/cheatcodes/prank.rs:24]: I wonder if we can make a general function just like you did 
with `get_cheat_for_contract` (really good idea btw). We then could pass different mutable references to the state for different cheatcodes (e.g. for `prank` we would pass `&mut self.pranked_contracts` and `&mut self.global_prank`).

File: CHANGELOG.md [hunk 2]
```diff
@@ -13,6 +13,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - `--color` option to control when colored output is used
 
+#### Changed
+- `start_warp` and `stop_warp` now take `CheatTarget` as the first argument instead of `ContractAddress`. Read more [here](./docs/src/appendix/cheatcodes/start_warp.md). 
+
```
Comment (war-in) [inline CHANGELOG.md:18]: we should also mention `start_prank` here

File: crates/forge/src/test_execution_syscall_handler.rs [hunk 3]
```diff
@@ -258,39 +258,47 @@ impl TestExecutionSyscallHandler<'_, '_> {
...
+
                 Ok(())
             }
             "start_prank" => {
-                let contract_address = inputs[0].to_contract_address();
-                let caller_address = inputs[1].to_contract_address();
+                // The last element in `inputs` should be the contract address in all cases
+                let caller_address = inputs.last().unwrap().to_contract_address();
+
+                let target = deserialize_cheat_target(&inputs[..inputs.len() - 1]);
```
Comment (war-in) [inline crates/forge/src/test_execution_syscall_handler.rs:287]: Let's change the order of these two (we should extract them the same way they were passed)

File: crates/forge/src/test_execution_syscall_handler.rs [hunk 4]
```diff
@@ -695,6 +703,22 @@ fn handle_deploy_result(
...
+        Some(0) => CheatTarget::All,
+        Some(1) => CheatTarget::One(inputs[1].to_contract_address()),
+        Some(2) => {
+            let contract_addresses: Vec<_> = inputs[2..]
+                .iter()
+                .map(Felt252::to_contract_address)
+                .collect();
+            CheatTarget::Multiple(contract_addresses)
+        }
+        _ => panic!("Invalid CheatTarget variant"),
```
Comment (war-in) [inline crates/forge/src/test_execution_syscall_handler.rs:718]: It is `unreachable!()`

File: crates/cheatnet/src/state.rs [hunk 5]
```diff
@@ -264,3 +295,66 @@ fn match_node_response<T: Default>(result: StateResult<T>) -> StateResult<T> {
...
+    } else {
+        global_cheat.clone()
+    }
+}
+
+pub fn start_cheat<T: Clone, S: BuildHasher>(
+    global_cheat: &mut Option<T>,
+    contract_cheats: &mut HashMap<ContractAddress, CheatStatus<T>, S>,
+    target: CheatTarget,
+    cheat_variable: T,
```
Comment (piotmag769) [inline crates/cheatnet/src/state.rs:318]: `cheat_value` is a better name prob

File: crates/forge/tests/e2e/running.rs [hunk 6]
```diff
@@ -633,7 +633,7 @@ fn init_new_project_test() {
 
     let remote_url = get_remote_url();
     let branch = get_current_branch();
-
+
```
Comment (piotmag769) [inline crates/forge/tests/e2e/running.rs:636]: unrelated

**Review by piotmag769 — APPROVED** [review]
Really like the abstraction of cheats!

File: CHANGELOG.md (hunk 2 above)
Comment (bllu404) [reply CHANGELOG.md:18]: I added an identical bullet point for `start_prank`, but once all the cheats use the new interface we should probably just consolidate them into 1 bullet point.

**Comment by Arcticae** [comment]
Closing in favor of #1117 1117
